File: backend/modules/Portfolio/CromaticaCreativa.Modules.Portfolio.Domain/Aggregates/CorporateClient.cs
using CromaticaCreativa.Modules.Portfolio.Domain.Enums;
using CromaticaCreativa.Modules.Portfolio.Domain.ValueObjects;

namespace CromaticaCreativa.Modules.Portfolio.Domain.Aggregates;

public sealed class CorporateClient
{
    private CorporateClient(
        CorporateClientId id,
        CorporateClientName name,
        MediaReference logo)
    {
        Id = id ?? throw new ArgumentNullException(nameof(id));
        Name = name ?? throw new ArgumentNullException(nameof(name));
        Logo = logo ?? throw new ArgumentNullException(nameof(logo));
        Visibility = VisibilityStatus.Hidden;
    }

    public CorporateClientId Id { get; }

    public CorporateClientName Name { get; private set; }

    public MediaReference Logo { get; private set; }

    public VisibilityStatus Visibility { get; private set; }

    public static CorporateClient Create(
        CorporateClientId id,
        CorporateClientName name,
        MediaReference logo) =>
        new(id, name, logo);

    public void Rename(CorporateClientName name) =>
        Name = name ?? throw new ArgumentNullException(nameof(name));

    public void ChangeLogo(MediaReference logo) =>
        Logo = logo ?? throw new ArgumentNullException(nameof(logo));

    public void Show() => Visibility = VisibilityStatus.Visible;

    public void Hide() => Visibility = VisibilityStatus.Hidden;
}

File: backend/src/modules/Portfolio/Portfolio.Domain/Aggregates/CorporateClient.ts
import { VisibilityStatus } from '../Enums/VisibilityStatus';
import { CorporateClientId } from '../ValueObjects/CorporateClientId';
import { CorporateClientName } from '../ValueObjects/CorporateClientName';
import { MediaReference } from '../ValueObjects/MediaReference';

export class CorporateClient {
  private constructor(
    public readonly id: CorporateClientId,
    private _name: CorporateClientName,
    private _logo: MediaReference,
    private _visibility: VisibilityStatus,
  ) {}

  public static create(id: CorporateClientId, name: CorporateClientName, logo: MediaReference): CorporateClient {
    return new CorporateClient(id, name, logo, VisibilityStatus.Hidden);
  }

  public get name(): CorporateClientName { return this._name; }
  public get logo(): MediaReference { return this._logo; }
  public get visibility(): VisibilityStatus { return this._visibility; }
  public rename(name: CorporateClientName): void { this._name = name; }
  public changeLogo(logo: MediaReference): void { this._logo = logo; }
  public show(): void { this._visibility = VisibilityStatus.Visible; }
  public hide(): void { this._visibility = VisibilityStatus.Hidden; }
}
